src/nvim/eval.c

```diff
@@ -10733,6 +10733,17 @@ static void f_has(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     n = true;
   }
 
+  if (STRICMP(name, "ruby") == 0 && n == true) {
+    char *rubyhost = call_func_retstr("provider#ruby#Detect", 0, NULL, true);
```

Comment:
Can't we add the check in `provider#ruby#Detect()`? Why add special handling in C?

Reply:
The point is to improve `has()`, so that it works like in Vim. (The same should be done for Python.) I don't see how to do that without adding C code.

Reply:
You mean that code should fail? Why shouldn't it fail if the gem is not installed?

Reply:
No, I meant it must fail if the gem is not installed. `has('ruby')` should only return 1 if the Ruby support is really working, not just indicating that the shipped Ruby provider is available.

Reply:
I must be confused, that _is_ how the python provider works.

Reply:
And the entire purpose of provider#ruby#Detect is to check for this..

Reply:
I did `pip3 uninstall neovim` and `has('python3')` still returns 1. I don't even have pip installed for Python 2, let alone the neovim module installed, but `has('python')` still returns 1.

Reply:
It checks for Python installations, not the module: https://github.com/neovim/neovim/blob/6710164c2c44335a916a728cc8eb329c69d155f4/runtime/autoload/provider/pythonx.vim#L40-L71

Reply:
Why can't it check for the module?

Reply:
It probably should. Currently `#Detect()` for Ruby and Python do different things. The former looks for an executable installed by the gem, the latter for Python installations.

IMHO, Python's `#Detect()` should report if the neovim module is installed or return an empty string. Then we could introduce a new C function other than `eval_has_provider()` (which only checks for the existence of `provider#<lang>#Detect()`), let's call it `eval_has_support()`. It should check the actual return value of `#Detect()` and should be used by `f_has()`.

Reply:
the pythonx Detect() function already returns a List with empty element 0, in case of error:

https://github.com/neovim/neovim/blob/6710164c2c44335a916a728cc8eb329c69d155f4/runtime/autoload/provider/pythonx.vim#L69

Can't we do something similar--without any new functions, or new C code--for the case where the module isn't installed? It's just an extra `if` check.

Reply:
We have no C code that ever runs the `#Detect()` functions. We have to unify the output of the `#Detect()` functions and put the approach this commit PR takes into a new C function. Hence the suggestion for `eval_has_support()` (check return value of `#Detect()`) akin to `eval_has_provider()` (check for existence of `#Detect()`).

Then we would change `f_has()` from..

```c
  if (!n && eval_has_provider(name)) {
    n = true;
  }

  if (STRICMP(name, "ruby") == 0 && n == true) {
    char *rubyhost = call_func_retstr("provider#ruby#Detect", 0, NULL, true);
    if (rubyhost) {
      if (*rubyhost == NUL) {
        // Invalid rubyhost executable. Gem is probably not installed.
        n = false;
      }
      xfree(rubyhost);
    }
  }
```

..to..

```c
  if (!n && eval_has_support(name)) {
    n = true;
  }

  if (!n && eval_has_provider(name)) {
    n = true;
  }
```

Afterwards no additional C code is needed anymore. Everything could be done in VimL.

Reply:
> no C code that ever runs the #Detect() functions

Oops, my mistake. We can just modify `eval_has_provider` (`CHECK_PROVIDER`) to do that, then.

Reply:
Okay, cool. I'll whip something up in the next days.